File: bpsk_Rayleigh.m
% MATLAB Script for computing the BER for BPSK modulation in a
% Rayleigh fading channel and compred to AWGN Channel
% Clear all the previously used variables
clear all;
close all
 
format long;
% Frame Length
bit_count = 2;
% Number of the frames
frames = 1;
%Range of SNR over which to simulate
SNR = [0 5 10 20 30 40 50];
% Start the main calculation loop
for aa = 1: 1: length(SNR)
    
    % Initiate variables
    T_Errors = 0;
    T_bits = 0;
    
    % Keep going until you get 100 errors
    for frame = 1:1:frames
    
        % Generate some random bits
        uncoded_bits  = round(rand(1,bit_count));
        % BPSK modulator
        tx = -2*(uncoded_bits-0.5);
    
        % Noise variance
        N0 = 1/10^(SNR(aa)/10); % noise over 1 bit data
        
        % Rayleigh channel fading
        h = 1/sqrt(2)*[randn(1,length(tx)) + j*randn(1,length(tx))] 
        % Send over Gaussian Link to the receiver
        rx = h.*tx + sqrt(N0/2)*(randn(1,length(tx))+i*randn(1,length(tx)));
        
%---------------------------------------------------------------
        % Equalization to remove fading effects. Ideal Equalization
        % Considered
        rx = rx./h;
        
        % BPSK demodulator at the Receiver
        rx2 = rx < 0; % less than 0 --> 0, greater than 0 --> 1 ??
    
        % Calculate Bit Errors
        diff = uncoded_bits - rx2;
        T_Errors = T_Errors + sum(abs(diff));
        T_bits = T_bits + length(uncoded_bits);
        
    end
    % Calculate Bit Error Rate
    BER(aa) = T_Errors / T_bits;
    disp(sprintf('bit error probability = %f',BER(aa)));
end
  
%------------------------------------------------------------
% Finally plot the BER Vs. SNR(dB) Curve on logarithmic scale 
% Calculate BER through Simulation
% Rayleigh Theoretical BER
SNRLin = 10.^(SNR/10);
theoryBer = 0.5.*(1-sqrt(SNRLin./(SNRLin+1)));
% Start Plotting
% Rayleigh Theoretical BER
figure(1);
semilogy(SNR,theoryBer,'-','LineWidth',2);
hold on;
% Simulated BER
figure(1);
semilogy(SNR,BER,'or','LineWidth',2);
hold on;
xlabel('SNR (dB)');
ylabel('BER');
title('SNR Vs BER plot for BPSK Modualtion in Rayleigh Channel');
% Theoretical BER
figure(1);
theoryBerAWGN = 0.5*erfc(sqrt(10.^(SNR/10)));
semilogy(SNR,theoryBerAWGN,'blad-','LineWidth',2);
legend('Rayleigh Theoretical','Rayleigh Simulated', 'AWGN Theoretical');
axis([0 40 10^-5 0.5]);
grid on;

File: bpsk_rayleigh.py

import numpy as np
import pandas as pd 
from numpy import random
import matplotlib.pyplot as plt
# import cmath



# the number of bits per frame, default 10000 bits per frame
bit_per_frame = 10000
# the number of frames
number_frame = 1000
# range of examined signal-to-noise ratio, in dB
SNR_array = np.array([0,5,10,20,30,40,50])
print(SNR_array)

BER_array = []

# a = np.random.rand(1,10)
# print(a)

for snr in SNR_array:
	# initialize the bit error counter per frame
	T_errors = 0
	# total bits transmitted
	T_bits = 0
	# print('check',snr)

	for i in range(number_frame):
		# print('check',i)
		uncoded_bits = np.round(random.rand(1,bit_per_frame))
		# print(uncoded_bits)
		tx_bpsk = 2*uncoded_bits - 1
		# print(tx_bpsk,len(tx_bpsk[0]))
		# noise variance
		N0 = 1/(10**(snr/10))
		# print(N0)
		# Rayleigh channel fading
		h = 1/np.sqrt(2)*(random.randn(1,bit_per_frame)+random.randn(1,bit_per_frame)*1j)
		# print(h)
		rx = tx_bpsk*h + np.sqrt(N0/2)*(random.randn(1,bit_per_frame)+random.randn(1,bit_per_frame)*1j)

		rx_remove_fading = rx/h
		# print(rx_remove_fading)

		rx_demodulated = (rx_remove_fading > 0)*1
		# print(rx_demodulated)

		diff = rx_demodulated - uncoded_bits
		# print(diff)
		T_errors = T_errors + np.sum(np.abs(diff))
		# print(T_errors)
		T_bits = T_bits + bit_per_frame
	BER_array.append(T_errors/T_bits)
	print('bit error probability =', T_errors/T_bits)

line_sim, = plt.scatter(SNR_array, BER_array, c='blue')


SNR_dec = 10**(SNR_array/10)
# print(SNR_dec)
theoryBer = 0.5*(1-np.sqrt(SNR_dec/(SNR_dec+1)))
# print(theoryBer)
line_theory, = plt.plot(SNR_array, theoryBer, c='red') # checked the calculation


plt.yscale('log')
plt.legend([line_sim, line_theory], ['Sim', 'Theory'])
plt.xlabel('SNR (dB)')
plt.ylabel('BER')
plt.grid(linewidth=0.2)



plt.show()




